<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LPWC: Lag Penalized Weighted Correlation for Clustering Short Time Series • LPWC</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LPWC: Lag Penalized Weighted Correlation for Clustering Short Time Series">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LPWC</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/LPWC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gitter-lab/LPWC">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>LPWC: Lag Penalized Weighted Correlation for Clustering Short Time Series</h1>
                        <h4 class="author">Thevaa Chandereng &amp; Anthony Gitter</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/gitter-lab/LPWC/blob/master/vignettes/LPWC.Rmd"><code>vignettes/LPWC.Rmd</code></a></small>
      <div class="hidden name"><code>LPWC.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Time series data are a series of observations collected across different timepoints in which the ordering of the observations can convey important information. In biology, most time series data consist of a limited number of time points, complicating the analysis. An example of biological time series data is gene expression data, where the relative abundances of all genes are recorded over multiple minutes, hours, days, or weeks.</p>
<p>One property of a good clustering method for time series data is that it groups similar temporal shapes. Similarity in expression patterns may correspond to similarity in biological function, which helps aid in the direction of future research. Lag Penalized Weighted Correlation (LPWC) is a distance-based clustering algorithm that builds upon weighted correlation. LPWC aligns time series data to accommodate lags, allowing two entities (for example, genes) that exhibit similar temporal behaviors that are not perfectly synchronized to be clustered together. Lags are penalized using a Gaussian kernel so that synchronized temporal patterns are preferred over lagged patterns. Unlike many other clustering approaches, LPWC also accounts for irregular time sampling in which there are non-uniform intervals between timepoints.</p>
<p>If you use LPWC in published research, please cite our preprint:</p>
<p>Lag Penalized Weighted Correlation for Time Series Clustering. Thevaa Chandereng, Anthony Gitter. bioRxiv 2018. <a href="https://doi.org/10.1101/292615">doi:10.1101/292615</a></p>
<p>If after reading through this vignette you have questions or problems using LPWC, please post them to <a href="https://github.com/gitter-lab/LPWC/issues" class="uri">https://github.com/gitter-lab/LPWC/issues</a>. This will notify the package maintainers and benefit other users.</p>
<p>You should <strong>not</strong> email your question to the package authors, as we will just reply that the question should be posted to the <strong>Github support site</strong>.</p>
</div>
<div id="running-lpwc" class="section level1">
<h1 class="hasAnchor">
<a href="#running-lpwc" class="anchor"></a>Running LPWC</h1>
<p>Prior to analyzing your data, the R package needs to be installed.</p>
<p>The easiest way to install LPWC is through CRAN:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"LPWC"</span>)</code></pre></div>
<p>There are other additional ways to download LPWC. The first option is most useful if want to download a specific version of LPWC (which can be found at <a href="https://github.com/gitter-lab/LPWC/releases" class="uri">https://github.com/gitter-lab/LPWC/releases</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"gitter-lab/LPWC@vx.xx.x"</span>)
<span class="co">#OR </span>
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_version">install_version</a></span>(<span class="st">"LPWC"</span>, <span class="dt">version =</span> <span class="st">"x.x.x"</span>, <span class="dt">repos =</span> <span class="st">"http://cran.us.r-project.org"</span>)</code></pre></div>
<p>The second option is to download through GitHub.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"gitter-lab/LPWC"</span>)</code></pre></div>
<p>After successful installation, the package must be loaded into the working space:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(LPWC)
<span class="co">#&gt; Warning: package 'LPWC' was built under R version 3.4.4</span></code></pre></div>
<div id="required-input" class="section level2">
<h2 class="hasAnchor">
<a href="#required-input" class="anchor"></a>Required input</h2>
<p><strong>Data</strong>: Input to LPWC is a matrix. We assume here that it is a gene expression matrix, but other data types are suitable as well. The expression matrix should be n -by- p where n is the number of genes (in rows) and p is the number of timepoints.</p>
<p>The object <strong>simdata</strong> is a simulated dataset for 200 genes with 8 timepoints that follow the biological <a href="https://doi.org/10.1093/bioinformatics/btw665">impulse model</a>. This is stored as a matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(simdata)
simdata[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, ]
<span class="co">#&gt;           Time_ 1   Time_ 2   Time_ 3   Time_ 4  Time_ 5  Time_ 6  Time_ 7</span>
<span class="co">#&gt; Gene_ 1 13.180777 17.125143 15.770959 16.541127 22.20278 29.60190 30.59476</span>
<span class="co">#&gt; Gene_ 2 24.022276 23.549481 20.717807 30.725417 39.31060 33.87521 36.15077</span>
<span class="co">#&gt; Gene_ 3 13.808949 15.585296 14.682402 22.498271 28.18709 28.53901 27.99949</span>
<span class="co">#&gt; Gene_ 4  9.527839  6.639082  8.841857  9.724606 18.98756 22.94803 24.47678</span>
<span class="co">#&gt; Gene_ 5 13.682422 10.835738 18.088811 20.666888 26.19007 25.95533 26.82196</span>
<span class="co">#&gt;          Time_ 8  Time_ 9 Time_ 10</span>
<span class="co">#&gt; Gene_ 1 31.43318 20.22925 18.89403</span>
<span class="co">#&gt; Gene_ 2 35.95496 29.87581 29.61810</span>
<span class="co">#&gt; Gene_ 3 34.25247 23.11201 26.29722</span>
<span class="co">#&gt; Gene_ 4 26.75902 19.61690 21.17882</span>
<span class="co">#&gt; Gene_ 5 24.18250 21.13418 19.65256</span>
<span class="kw">str</span>(simdata)
<span class="co">#&gt; 'data.frame':    200 obs. of  10 variables:</span>
<span class="co">#&gt;  $ Time_ 1 : num  13.18 24.02 13.81 9.53 13.68 ...</span>
<span class="co">#&gt;  $ Time_ 2 : num  17.13 23.55 15.59 6.64 10.84 ...</span>
<span class="co">#&gt;  $ Time_ 3 : num  15.77 20.72 14.68 8.84 18.09 ...</span>
<span class="co">#&gt;  $ Time_ 4 : num  16.54 30.73 22.5 9.72 20.67 ...</span>
<span class="co">#&gt;  $ Time_ 5 : num  22.2 39.3 28.2 19 26.2 ...</span>
<span class="co">#&gt;  $ Time_ 6 : num  29.6 33.9 28.5 22.9 26 ...</span>
<span class="co">#&gt;  $ Time_ 7 : num  30.6 36.2 28 24.5 26.8 ...</span>
<span class="co">#&gt;  $ Time_ 8 : num  31.4 36 34.3 26.8 24.2 ...</span>
<span class="co">#&gt;  $ Time_ 9 : num  20.2 29.9 23.1 19.6 21.1 ...</span>
<span class="co">#&gt;  $ Time_ 10: num  18.9 29.6 26.3 21.2 19.7 ...</span></code></pre></div>
<p><strong>Timepoints:</strong> The object <strong>timepoints</strong> should be a vector of timepoints that specify when the data were collected. This should match the column length in <strong>Data</strong>. This can be in any form (seconds, minutes, hours, or even days), but make sure the units are uniform throughout.</p>
<p>The timepoints used in the example <strong>simdata</strong> are</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">timepoints &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">18</span>, <span class="dv">24</span>, <span class="dv">32</span>, <span class="dv">48</span>, <span class="dv">72</span>)
timepoints
<span class="co">#&gt;  [1]  0  2  4  6  8 18 24 32 48 72</span></code></pre></div>
<p><strong>Data</strong> and <strong>Timepoints</strong> are the main requirement to run LPWC, however, there are other optional parameters that affect the results. Our manuscript referenced above describes these parameters in more detail.</p>
<p><strong>Max.lag:</strong> The object <strong>max.lag</strong> should be an integer value that controls the maximum lag. This max.lag has to be less than the floor of length(timepoints) / 4. That is because longer lags lead to comparisons between two expression profiles that use less than half of the data points, which can lead to spurious correlations.</p>
<p><strong>C:</strong> The object <strong>C</strong> should be a numeric. This controls the width of the Gaussian kernel function used to penalize the lag and the weights in weighed correlation vector. The parameter <strong>C</strong> can be set automatically by setting <strong>C</strong> to the default NULL value and setting the <strong>penalty</strong> argument described below. If <strong>C</strong> is set to a specific value, the <strong>penalty</strong> is ignored.</p>
<p><strong>penalty:</strong> The object <strong>penalty</strong> only allows two values: “low” and “high”. It is used to automatically set the value of <strong>C</strong>. High imposes a higher penalty on lags thus favoring grouping genes without introducing lags. The lower penalty leads to more lagged genes in the clusters. The default is set to the “high” penalty. This argument is ignored if <strong>C</strong> is set to a specific value.</p>
<p><strong>iter:</strong> The object <strong>iter</strong> should be a numeric. This controls the number of values of C that are tested when automatically setting the C parameter. The default is 10. Increasing iter increases computational time. This is only relevant for the low penalty because the high penalty uses a single C instead of searching over multiple C.</p>
</div>
<div id="interpreting-lags" class="section level2">
<h2 class="hasAnchor">
<a href="#interpreting-lags" class="anchor"></a>Interpreting lags</h2>
<p>Lags are taken with respect to the original timepoints vector. Lags can be positive or negative integers and specify the number of indices to shift the timepoints. The examples below demonstrate applying different lags to genes 1 and 2.</p>
<p><img src="LPWC_files/figure-html/unnamed-chunk-4-1.png" width="384" style="display: block; margin: auto;"></p>
<table class="table"><tbody>
<tr class="odd">
<td align="left">Original</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
<tr class="even">
<td align="left">Gene1</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
<tr class="odd">
<td align="left">Gene2</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
</tbody></table>
<p><img src="LPWC_files/figure-html/unnamed-chunk-5-1.png" width="384" style="display: block; margin: auto;"></p>
<table class="table"><tbody>
<tr class="odd">
<td align="left">Original</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
<tr class="even">
<td align="left">Gene1</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
<td align="center">-</td>
</tr>
<tr class="odd">
<td align="left">Gene2</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
</tbody></table>
<p><img src="LPWC_files/figure-html/unnamed-chunk-6-1.png" width="384" style="display: block; margin: auto;"></p>
<table class="table"><tbody>
<tr class="odd">
<td align="left">Original</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
</tr>
<tr class="even">
<td align="left">Gene1</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
<td align="center">75</td>
<td align="center">-</td>
</tr>
<tr class="odd">
<td align="left">Gene2</td>
<td align="center">-</td>
<td align="center">0</td>
<td align="center">5</td>
<td align="center">15</td>
<td align="center">30</td>
<td align="center">45</td>
<td align="center">60</td>
</tr>
</tbody></table>
<p>The time series plots show aligned expression vectors, and the tables show the aligned timepoints. With no lags, the temporal profiles of genes 1 and 2 are not aligned so the gene pair will have a low LPWC similarity score (top row). When gene 1 has a lag of -1, the patterns are aligned, and the LPWC similarity score will be high (middle row). Finally, when gene 1 has a lag of -1 and gene 2 has a lag of 1, the temporal shapes are once again not aligned. In this case, the LPWC similarity score will be even lower than in the no lag case because the penalty for introducing lags is applied (bottom row).</p>
</div>
<div id="obtaining-gene-gene-similarities" class="section level2">
<h2 class="hasAnchor">
<a href="#obtaining-gene-gene-similarities" class="anchor"></a>Obtaining gene-gene similarities</h2>
<p>LPWC computes similarity (an adjusted weighted correlation) between genes and returns three outputs: correlation of genes in dist form (see as.dist), the best lag used for each gene, and the C used. The lags introduced are in reference to the original time points (i.e. lag of 0).</p>
<p>An example output of LPWC for 10 genes using high penalty follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">LPWC<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/LPWC/topics/corr.bestlag">corr.bestlag</a></span>(simdata[<span class="dv">49</span><span class="op">:</span><span class="dv">58</span>, ], <span class="dt">timepoints =</span> timepoints, <span class="dt">max.lag =</span> <span class="dv">2</span>, <span class="dt">penalty =</span> <span class="st">"high"</span>, <span class="dt">iter =</span> <span class="dv">10</span>)
<span class="co">#&gt; $corr</span>
<span class="co">#&gt;               1            2            3            4            5</span>
<span class="co">#&gt; 2   0.851923851                                                    </span>
<span class="co">#&gt; 3  -0.218237490  0.185296315                                       </span>
<span class="co">#&gt; 4  -0.503013415 -0.185206626  0.820661318                          </span>
<span class="co">#&gt; 5  -0.459664840 -0.098109602  0.846568127  0.902530166             </span>
<span class="co">#&gt; 6  -0.299615918  0.147669259  0.942905074  0.707652473  0.725611385</span>
<span class="co">#&gt; 7  -0.324090620  0.008691392  0.801844343  0.827065718  0.783988169</span>
<span class="co">#&gt; 8  -0.311218630  0.145879675  0.958900623  0.806729108  0.821231416</span>
<span class="co">#&gt; 9  -0.177943816  0.183155731  0.818514482  0.584610971  0.733599647</span>
<span class="co">#&gt; 10 -0.412958564  0.050367297  0.943102621  0.773563872  0.818502367</span>
<span class="co">#&gt;               6            7            8            9</span>
<span class="co">#&gt; 2                                                     </span>
<span class="co">#&gt; 3                                                     </span>
<span class="co">#&gt; 4                                                     </span>
<span class="co">#&gt; 5                                                     </span>
<span class="co">#&gt; 6                                                     </span>
<span class="co">#&gt; 7   0.686430019                                       </span>
<span class="co">#&gt; 8   0.961429975  0.712528238                          </span>
<span class="co">#&gt; 9   0.803809413  0.756633891  0.740717669             </span>
<span class="co">#&gt; 10  0.951300777  0.727622247  0.937457260  0.748985339</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $lags</span>
<span class="co">#&gt;  [1] 0 1 0 0 0 0 0 0 0 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $C</span>
<span class="co">#&gt; [1] 306.2401</span></code></pre></div>
<p>However, many clustering algorithms, such as hierarchical clustering, take distances as input instead of similarities. We convert the similarities to distances by subtracting the similarities from 1. Initially</p>
<p><span class="math display">\[ -1 \leq Corr \leq 1\]</span></p>
<p>When we take Dist = 1 - Corr</p>
<p><span class="math display">\[ 0 \leq Dist = (1 - Corr) \leq 2\]</span></p>
</div>
<div id="generating-clusters-using-hierarchical-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-clusters-using-hierarchical-clustering" class="anchor"></a>Generating clusters using hierarchical clustering</h2>
<p>LPWC computes a similarity matrix and does not directly cluster the genes. A standard similarity-based clustering algorithm such as hierarchical clustering can be applied to the LPWC similarities.</p>
<p>An example for 10 genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dist &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>LPWC<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/LPWC/topics/corr.bestlag">corr.bestlag</a></span>(simdata[<span class="dv">11</span><span class="op">:</span><span class="dv">20</span>, ], <span class="dt">timepoints =</span> timepoints, <span class="dt">max.lag =</span> <span class="dv">2</span>, <span class="dt">penalty =</span> <span class="st">"low"</span>, <span class="dt">iter =</span> <span class="dv">10</span>)<span class="op">$</span>corr
<span class="kw">plot</span>(<span class="kw">hclust</span>(dist))</code></pre></div>
<p><img src="LPWC_files/figure-html/clust1-1.png" width="480"></p>
<p>The genes can also be assigned to clusters using cutree function. An example of cluster assignment with 3 clusters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dist &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>LPWC<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/LPWC/topics/corr.bestlag">corr.bestlag</a></span>(simdata[<span class="dv">11</span><span class="op">:</span><span class="dv">20</span>, ], <span class="dt">timepoints =</span> timepoints, <span class="dt">max.lag =</span> <span class="dv">2</span>, <span class="dt">penalty =</span> <span class="st">"low"</span>, <span class="dt">iter =</span> <span class="dv">10</span>)<span class="op">$</span>corr
<span class="kw">cutree</span>(<span class="kw">hclust</span>(dist), <span class="dt">k =</span> <span class="dv">3</span>)
<span class="co">#&gt;  [1] 1 2 3 2 3 3 2 2 3 1</span></code></pre></div>
</div>
</div>
<div id="contributors" class="section level1">
<h1 class="hasAnchor">
<a href="#contributors" class="anchor"></a>Contributors</h1>
<p>Our development of LPWC benefited from help and feedback from many individuals, including but not limited to:</p>
<p>Wenzhi Cao, Karl Broman, Jen Birstler, James Dowell, Ron Stewart, leungi (Github user), Gitter Lab members.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.4.3 (2017-11-30)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: OS X El Capitan 10.11.4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.0.0.9000 LPWC_0.99.3       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_0.12.17     highr_0.6        compiler_3.4.3   pillar_1.2.2    </span>
<span class="co">#&gt;  [5] plyr_1.8.4       bindr_0.1.1      tools_3.4.3      digest_0.6.15   </span>
<span class="co">#&gt;  [9] evaluate_0.10.1  memoise_1.1.0    tibble_1.4.2     gtable_0.2.0    </span>
<span class="co">#&gt; [13] pkgconfig_2.0.1  rlang_0.2.1      rstudioapi_0.7   commonmark_1.5  </span>
<span class="co">#&gt; [17] yaml_2.1.19      pkgdown_1.1.0    bindrcpp_0.2.2   withr_2.1.2     </span>
<span class="co">#&gt; [21] stringr_1.3.1    dplyr_0.7.6      roxygen2_6.0.1   xml2_1.2.0      </span>
<span class="co">#&gt; [25] knitr_1.20       desc_1.2.0       fs_1.2.2         rprojroot_1.3-2 </span>
<span class="co">#&gt; [29] grid_3.4.3       tidyselect_0.2.4 glue_1.2.0       R6_2.2.2        </span>
<span class="co">#&gt; [33] rmarkdown_1.9    purrr_0.2.5      magrittr_1.5     backports_1.1.2 </span>
<span class="co">#&gt; [37] scales_0.5.0     htmltools_0.3.6  MASS_7.3-50      assertthat_0.2.0</span>
<span class="co">#&gt; [41] colorspace_1.3-2 labeling_0.3     stringi_1.2.2    lazyeval_0.2.1  </span>
<span class="co">#&gt; [45] munsell_0.4.3    crayon_1.3.4</span></code></pre></div>
</div>
<div id="faq" class="section level1">
<h1 class="hasAnchor">
<a href="#faq" class="anchor"></a>FAQ</h1>
<ul>
<li>How can I speed up LPWC for the low penalty with a large dataset?</li>
</ul>
<p>We recommend parallelizing parts of the analysis, as shown below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">

<span class="co"># This function stores two different list separately </span>
comb &lt;-<span class="st"> </span><span class="cf">function</span>(x, ...) {
  <span class="kw">lapply</span>(<span class="kw">seq_along</span>(x),
         <span class="cf">function</span>(i) <span class="kw">c</span>(x[[i]], <span class="kw">lapply</span>(<span class="kw">list</span>(...), <span class="cf">function</span>(y) y[[i]])))
}


<span class="co"># adding the data</span>
data &lt;-<span class="st"> </span>simdata[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]
<span class="co"># number of iterations</span>
iter &lt;-<span class="st"> </span><span class="dv">10</span>
<span class="co"># C values that are used in the algorithm</span>
allC &lt;-<span class="st"> </span><span class="kw"><a href="../reference/findC.html">findC</a></span>(<span class="dt">timepoints =</span> timepoints, <span class="dt">iter =</span> iter)

<span class="co"># setting the clusters </span>
core &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/parallel/topics/detectCores">detectCores</a></span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>
cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/parallel/topics/makeCluster">makeCluster</a></span>(core)

<span class="co"># assigning the parallelization</span>
doParallel<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/doParallel/topics/registerDoParallel">registerDoParallel</a></span>(cl)


## running the algorithm for different C 
result &lt;-<span class="st"> </span><span class="kw">foreach</span>(<span class="dt">i =</span> <span class="dv">1</span><span class="op">:</span>iter, <span class="dt">.combine=</span><span class="st">'comb'</span>, <span class="dt">.multicombine=</span><span class="ot">TRUE</span>,
                <span class="dt">.init=</span><span class="kw">list</span>(<span class="kw">list</span>(), <span class="kw">list</span>())) <span class="op">%dopar%</span><span class="st"> </span>{
            lags &lt;-<span class="st"> </span><span class="kw"><a href="../reference/best.lag.html">best.lag</a></span>(data, <span class="dt">max.lag =</span> <span class="dv">3</span>, <span class="dt">timepoints =</span> timepoints, <span class="dt">C =</span> allC[i])
            new.data &lt;-<span class="st"> </span><span class="kw"><a href="../reference/prep.data.html">prep.data</a></span>(<span class="dt">data =</span> data, <span class="dt">lags =</span> lags, <span class="dt">timepoints =</span> timepoints)
            corr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/comp.corr.html">comp.corr</a></span>(new.data<span class="op">$</span>data, new.data<span class="op">$</span>time, <span class="dt">C =</span> allC[i])
            <span class="kw">return</span>(<span class="kw">list</span>(corr, lags))
                }


<span class="co"># dividing the list into two different list: one for lags and one for all the correlations</span>
allcorr &lt;-<span class="st"> </span>result[[<span class="dv">1</span>]]
alllags &lt;-<span class="st"> </span>result[[<span class="dv">2</span>]]

<span class="co"># picking best C</span>
val &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, (<span class="kw">length</span>(iter) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>))
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(iter <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)){
  val[i] &lt;-<span class="st"> </span><span class="kw">sum</span>((<span class="kw">as.vector</span>(allcorr[[i <span class="op">+</span><span class="st"> </span><span class="dv">1</span>]]) <span class="op">-</span><span class="st"> </span><span class="kw">as.vector</span>(allcorr[[i]]))<span class="op">^</span><span class="dv">2</span>)
}

<span class="co"># returning the results for the best C</span>
result &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">corr =</span> allcorr[[<span class="kw">which.min</span>(val) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>]], <span class="dt">lags =</span> alllags[[<span class="kw">which.min</span>(val) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>]], <span class="dt">C =</span> values[<span class="kw">which.min</span>(val) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>])</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#running-lpwc">Running LPWC</a><ul class="nav nav-pills nav-stacked">
<li><a href="#required-input">Required input</a></li>
      <li><a href="#interpreting-lags">Interpreting lags</a></li>
      <li><a href="#obtaining-gene-gene-similarities">Obtaining gene-gene similarities</a></li>
      <li><a href="#generating-clusters-using-hierarchical-clustering">Generating clusters using hierarchical clustering</a></li>
      </ul>
</li>
      <li><a href="#contributors">Contributors</a></li>
      <li><a href="#session-info">Session info</a></li>
      <li><a href="#faq">FAQ</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Thevaa Chandereng, Anthony Gitter.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
